### ISSUE TITLE
vmap on location?

### ISSUE THREAD

**Issue by VENOM314** [jaxleyverse/jaxley#644]
Hello Jaxley community:

I am wondering if vmap can sweep across locations. I have tried one-hot encoding and amplitude sweep (only the site of interest receives non-zero amplitude) but I am setting a lot of data_stimuli so I am wondering if that might degrade memory usage (less batch size maybe?)

I understand it is technically difficult to implement a vmap on location: probably the reason is indexing in a dataframe needs _concrete variable_ as index, but vmap/jit makes variables _tracers_? Would there be other datatypes that can allow tracer index and solve this problem?

Also for setting ephys parameters, we have cell.set() and cell.data_set(). Is the key difference between these two pure and impure functions like this? https://docs.jax.dev/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions. I really want to learn more about the compilation behavior of Jax.

**Comment by michaeldeistler** [comment]
Hi! Indeed, you cannot `vmap` across locations. In principle, I think that it might be possible (although I am not yet fully sure), but the current version of `Jaxley` does not support it.

`cell.set()` writes the result to `cell.nodes`, which is a pandas Dataframe. This moves parameters to CPU, and it interrupts their gradient. In contrast `pstate = cell.data_set()` writes the parameters to the `pstate`, which is a dictionary of JAX arrays. Because of this, it can be jitted and differentiated through.

Hope this helps!
Michael

**Comment by michaeldeistler** [comment]
> I think the bottleneck might be that cell.nodes is a dataframe and cannot accept tracers as index, thus not working within jit/vmap.

Yes, I agree. @jnsbck is working on a new backend for `Jaxley`, I think this might solve this issue.

**Comment by VENOM314** [comment]
Hi all, I wanted to check back in and as I have some free time right now, would like to know if there is anywhere I can help. Would it be possible for me to get a pull request and try to work on this issue?

**Comment by michaeldeistler** [comment]
Yes, exactly! And yes, please keep us updated!

Michael

**Comment by michaeldeistler** [comment]
yes, that's exactly right.
